### ISSUE TITLE
OCI sources give wrong mediatype

### ISSUE THREAD

**Issue by cyphar** [projectatomic/skopeo#305]
This is caused by the change in `v1.0.0-rc4` of the spec, where the `mediaType` property of all objects was dropped. As a result, attempting to do something like this will fail:

```
% skopeo copy oci:someimage:latest docker-daemon:someimage:latest
Getting image source signatures
Copying blob sha256:5210fb0f39cf518766e18c208f70c2ce0a8eca7594c6607bd3661f2ba56a11b2
 44.75 MB / 47.03 MB [======================================================>--]
Copying config sha256:566ed144547ffb7c01ce10eddaf51d4a3e0f822cf169d1e8d2928f1de726911c
 0 B / 302 B [-----------------------------------------------------------------]
Writing manifest to image destination
FATA[0000] Error writing manifest: Unsupported manifest type, need a Docker schema 2 manifest
```

The reason for this being that `manifest.GuessMIMEType` will always return `application/vnd.docker.distribution.manifest.v2+json`. Really I'm wondering why we're using that function for individual transports -- surely the **transport** knows what mediatype a manifest is?

**Comment by cyphar** [comment]
In particular this was broken by opencontainers/image-spec#411. This is also currently breaking SUSE/kiwi#216.

I am getting **increasingly distressed** about the **lack of testing** people are doing within the spec writing community...

**Comment by runcom** [comment]
> Really I'm wondering why we're using that function for individual transports -- surely the transport knows what mediatype a manifest is?

if it doesn't make sense to use it, we can drop it I guess (you noted a transport knows what a mediatype a manifest is)

@mtrmac wdyt?

**Comment by cyphar** [comment]
At the very least, for OCI, we can **only** know the mediaType from `refs/...` (and soon `index.json`). So for OCI I'm going to change this but the general idea is a bit iffy in my view.

**Comment by mtrmac** [comment]
The transports _should_ return the true manifest type if they know it; and they are, per `types.go`, allowed to return `""` when unknown. The OCI situation is probably a copy&paste error from `dir:`, which does not have a MIME type indication. So, this particular situation seems to be just a bug, thanks for filing https://github.com/containers/image/issues/237 .

A bit more generally, `dir:` more or less has to guess because `image.manifestInstanceFromBlob` treats any unrecognized value (including the documented `""`= unknown) as docker schema1. Perhaps we need to change the API so that the MIME type returned from `GetManifest`, can differentiate “we know for certain” (OCI, docker/distribution registries when working), “we should know for certain but with this image we don’t” (e.g. missing `Content-Type` in HTTP headers in an old/dubious docker/distribution server implementation), and “we have no idea, please guess” (`dir:`). Or just move the “unrecognized/unknown/empty value means docker schema1” logic into `dockerImageSource`, it does seem specific to the docker/distribution transport semantics and not something so inherently general that it should live in `containers/image/image`.

---

Related, but also pretty much orthogonal, is that it would be nice if `GuessManifestMIMEType` were able to correctly differentiate between OCI and docker schema2. OTOH there is no obviously unique identifier for them AFAICS… Perhaps we could guess OCI instead of schema2 in the `schemaVersion == 2 && mediaType is unset` case; or we could see whether `config.mediaType` is an OCI type or a docker type (which seems more specific but may be technically incorrect, schema2 manifests may be able to carry OCI configs, if not now, then in the future).

Is it known why OCI removed the MIME type? Do they recommend any other replacement to help guessing / help detect incorrect MIME metadata?

**Comment by cyphar** [comment]
@mtrmac 

> Is it known why OCI removed the MIME type? Do they recommend any other replacement to help guessing / help detect incorrect MIME metadata?

Their logic for removing it is that all of the type information is stored in descriptors and that self-describing JSON blobs are the devil's handiwork. In all seriousness, I spent several days of my life trying to argue against the change in the thread because stuff like this would happen.

In effect their "recommendation" is to always access OCI images through descriptors, because that's the only "guaranteed way" of getting the right media type. Which is great fun.

**Comment by cyphar** [comment]
This was fixed by #306 which also added tests for this case.
